<!DOCTYPE html>
<html>
<head>
	<title>JS Errors</title>
</head>
<body>
	<h1>JavaScript Errors - Throw and Try to Catch</h1>
	<h2>Important Statements</h2>
		<ul>
			<li><strong>try-</strong> test a block of code for erros</li>
			<li><strong>catch-</strong> handle the errors</li>
			<li><strong>throw-</strong> create custom errors</li>
			<li><strong>finally-</strong> execute code after try and catch, regardless of the result</li>
		</ul>

		<h3>Errors Example</h3>
		<div id="ex1"></div>

		<h2>JavaScript try and catch</h2>
		<p><strong>Try</strong> statement allows you to define a block of code to be tested for errors while it is being executed.</p>
		<p><strong>Catch</strong> statement allows you to define a block of code to be executed, if an error occurs in the try block.</p>
		<p>Try & Catch come in pairs.</p>

		<h2>The throw Statement</h2>
		<p><strong>Throw</strong> statement allows you to create a custom error.</p>
		<p>Technically you can <strong>throw an exception (throw an error)</strong>.</p>
		<p>The exception can ba a JavaScript String, a Number, a Boolean or an Object.</p>

		<h3>Input Validation Example</h3>
		<p>Please input a number between 5 and 10:</p>
		<input id="demo" type="text">
		<button type="button" onclick="myFunction()">Test Input</button>
		<p id="message"></p>

		<h2>The finally Statement</h2>
		<p>The <strong>finally</strong> statement lets you execute code, after try and catch, regardless of the result:</p>
		<div id="finally"></div>

		<img src="errors.jpg">

		<h2>Eval Error</h2>
		<p>Indicates an error in the eval() function</p>
		<p>Newer versions of JS do not throw any EvalError. Use SyntaxError instead.</p>

		<h2>Range Error</h2>
		<p>A <strong>RangeError</strong> is thrown if you use a number that is outside the range of legal values.</p>
		<p>Example: num.toPrecision(500);</p>
		<div id="rangerr"></div>


	<script>
		
		try {
			addlert("Welcome guest!");
		}
		catch(err) {
			document.getElementById("ex1").innerHTML = err.message; 
		}


		// Input Validation
		function myFunction() {
			var messge, x;
			message = document.getElementById("message"); 
			message.innerHTML = "";
			x = document.getElementById("demo").value;
			try {
				if(x == "") throw "empty.";
				if (isNaN(x)) throw "not a number.";
				x = Number(x);
				if(x < 5) throw "too low.";
				if (x > 10) throw "too high.";
			}
			catch (err) {
				message.innerHTML = "Input is " + err; 
			}
		}


		// Finally Statement
		// function myFunction () {
		// 	var message2, x;
		// 	message2 = document.getElementById("finally");
		// 	message2.innerHTML = "";
		// 	x = document.getElementById("finally").value;
		// 	try {
		// 		if(x == "") throw "is empty";
		// 		if (isNaN(x)) throw "is not a number";
		// 		x = Number(x);
		// 		if(x > 10) throw "is too high";
		// 		if (x < 5) throw "is too low"; 
		// 	}
		// 	catch(err) {
		// 		message2.innerHTML = "Error: " + err + ".";
		// 	}
		// 	finally {
		// 		document.getElementById("finally").value = "";
		// 	}
		// }


		// Range Error
		var num = 1;
		try {
			num.toPrecision(500);  // A number cannot have 500 significant digits
		}
		catch(err) {
			document.getElementById("rangerr").innerHTML = err.name; 
		}

	</script>

</body>
</html>